File: admin/src/main/java/com/project/admin/service/TechBlogService.java
package com.project.admin.service;

import com.project.admin.constant.Status;
import com.project.admin.controller.dto.request.techblog.TechBlogListRequest;
import com.project.admin.domain.entity.TechBlog;
import com.project.admin.domain.entity.TechBlogApproval;
import com.project.admin.domain.exception.techblog.TechBlogErrorCode;
import com.project.admin.domain.exception.techblog.TechBlogException;
import com.project.admin.repository.TechBlogApprovalRepository;
import com.project.admin.repository.TechBlogRepository;
import com.project.admin.service.dto.techblog.TechBlogDetailResult;
import com.project.admin.service.dto.techblog.TechBlogResult;
import java.util.Map;
import java.util.stream.Collectors;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import java.util.List;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class TechBlogService {

    private final TechBlogRepository techBlogRepository;
    private final TechBlogApprovalRepository techBlogApprovalRepository;

    public Page<TechBlogResult> getTechBlogs(TechBlogListRequest request) {
        Pageable pageable = PageRequest.of(
                request.page() - 1,
                request.size(),
                Sort.Direction.DESC,
                "publishedAt");

        return techBlogRepository.findByStatusAndSource(request.status(), request.source(),
                        pageable)
                .map(TechBlogResult::from);
    }

    public TechBlogDetailResult getTechBlog(Long id) {
        TechBlog techBlog = techBlogRepository.findById(id)
                .orElseThrow(() -> new TechBlogException(TechBlogErrorCode.TECH_BLOG_ERROR_CODE));

        TechBlogApproval approval = techBlogApprovalRepository.findByTechBlogId(techBlog.getId())
                .orElse(null);

        return TechBlogDetailResult.from(techBlog, approval);
    }

    @Transactional
    public void updateStatus(List<Long> ids, Status status) {
        techBlogRepository.updateStatusByIds(ids, status);

        List<TechBlog> techBlogs = techBlogRepository.findAllById(ids);

        Map<Long, TechBlogApproval> approvalMap = techBlogApprovalRepository.findAllByTechBlogIdIn(
                        ids)
                .stream()
                .collect(Collectors.toMap(
                        a -> a.getTechBlog().getId(),
                        a -> a
                ));

        List<TechBlogApproval> approvals = techBlogs.stream()
                .map(techBlog -> approvalMap.getOrDefault(
                        techBlog.getId(),
                        new TechBlogApproval(techBlog)
                )).toList();

        techBlogApprovalRepository.saveAll(approvals);
    }
}

File: api/src/main/kotlin/com/project/api/service/TechBlogService.kt
package com.project.api.service

import com.project.api.constants.Region
import com.project.api.constants.Source
import com.project.api.exception.CommonErrorCode
import com.project.api.exception.QuickStackException
import com.project.api.repository.TechBlogRepository
import com.project.api.service.dto.CompanyResult
import com.project.api.service.dto.CompanyResults
import com.project.api.service.dto.PageResult
import com.project.api.service.dto.TechBlogDetailResult
import com.project.api.service.dto.TechBlogResult
import com.project.api.service.dto.TechBlogResults
import org.springframework.data.domain.PageRequest
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class TechBlogService(
    private val techBlogRepository: TechBlogRepository,
) {

    fun getCompanies(region: Region? = null): CompanyResults {
        val statsBySource = techBlogRepository.findCompanies()
            .associateBy { it.source }

        val companies = Source.entries
            .filter { region == null || it.region == region }
            .map { source -> CompanyResult.of(source, statsBySource[source]) }

        return CompanyResults(companies = companies)
    }

    fun getPopularCompanies(): CompanyResults {
        val companies = techBlogRepository.findPopularCompanies(Source.popularEntries)
            .map { stat -> CompanyResult.from(stat) }
        return CompanyResults(companies = companies)
    }

    fun getTechBlogs(keyword: String?, page: Int, size: Int): TechBlogResults {
        val pageable = PageRequest.of(page - 1, size)
        val result = techBlogRepository.findByKeyword(keyword?.takeIf { it.isNotBlank() }, pageable)
        return TechBlogResults(
            techBlogs = result.content.map { TechBlogResult.from(it) },
            pagination = PageResult.from(result),
        )
    }

    fun getTechBlogsBySource(sourceCode: String, page: Int, size: Int): TechBlogResults {
        val pageable = PageRequest.of(page - 1, size)
        val source = Source.fromValue(sourceCode) ?: return TechBlogResults.empty(page, size)
        val result = techBlogRepository.findBySource(source, pageable)
        return TechBlogResults(
            techBlogs = result.content.map { TechBlogResult.from(it) },
            pagination = PageResult.from(result),
        )
    }

    //상세 조회
    fun getTechBlog(id: Long): TechBlogDetailResult {
        val techBlog = techBlogRepository.findByIdAndPUBLISHED(id) ?: throw QuickStackException(
            CommonErrorCode.INVALID_REQUEST)
        return TechBlogDetailResult.from(techBlog);
    }
}
